Performing initial data inspection and cleaning of the UiS4ADL and PAAL ADL datasets.

In [1]:
import os
import warnings
import pandas as pd

# Suppress warnings from Python and external libraries
os.environ["PYTHONWARNINGS"] = "ignore"  # Suppress Python warnings
warnings.filterwarnings("ignore")

# UiS4ADL

In [2]:
fs_UiS4ADL = 100 # sampling frequency in Hz
path_UiS4ADL = "../Data/UiS4ADL/Raw/UiS4ADL_100hz_raw.csv"

df_UiS4ADL = pd.read_csv(path_UiS4ADL)

In [3]:
df_UiS4ADL

,timestamp,accX[mg],accY[mg],accZ[mg],gyroX[mdps],gyroY[mdps],gyroZ[mdps],magnX[mG],magnY[mG],magnZ[mG],...,fxGravityY[g],fxGravityZ[g],fxLinAccX[g],fxLinAccY[g],fxLinAccZ[g],fxHeading[deg],adl,session,subject,fileID
0,0.0,0.0,0.0,1009.0,490.0,350.0,-770.0,-289.0,4.0,-430.0,...,0.000,-1.000,0.001,0.000,0.009,181.065,0,1,1125,0
1,10.0,-1.0,1.0,1010.0,490.0,350.0,-770.0,-279.0,10.0,-439.0,...,-0.000,-1.000,0.001,0.001,0.010,181.065,0,1,1125,0
2,20.0,0.0,1.0,1011.0,420.0,280.0,-770.0,-289.0,0.0,-435.0,...,-0.000,-1.000,0.001,-0.000,0.011,181.065,0,1,1125,0
3,30.0,0.0,1.0,1009.0,350.0,280.0,-770.0,-286.0,12.0,-435.0,...,-0.000,-1.000,0.001,-0.000,0.009,181.065,0,1,1125,0
4,40.0,1.0,1.0,1010.0,420.0,280.0,-840.0,-282.0,0.0,-438.0,...,0.000,-1.000,0.001,-0.001,0.010,181.065,0,1,1125,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10559903,27730.0,836.0,-412.0,47.0,10640.0,10570.0,2590.0,-14.0,-103.0,-87.0,...,0.886,-0.093,0.043,0.050,-0.046,257.834,9,5,3245,4947
10559904,27740.0,827.0,-412.0,47.0,11760.0,10080.0,1610.0,-14.0,-103.0,-87.0,...,0.885,-0.096,0.043,0.058,-0.049,257.704,9,5,3245,4947
10559905,27750.0,816.0,-409.0,45.0,12740.0,9520.0,560.0,-12.0,-106.0,-84.0,...,0.885,-0.099,0.046,0.069,-0.054,257.563,9,5,3245,4947
10559906,27760.0,809.0,-410.0,42.0,13580.0,8820.0,-560.0,-17.0,-96.0,-84.0,...,0.886,-0.095,0.044,0.077,-0.053,257.724,9,5,3245,4947


## Basic data cleaning and overview

In [4]:
df_UiS4ADL = df_UiS4ADL[df_UiS4ADL['adl'] != 0] # removing the 0 class

print(f'The dataset contains {df_UiS4ADL["adl"].nunique()} unique ADL classes after removing the 0 class.')

The dataset contains 24 unique ADL classes after removing the 0 class.


In [5]:
df_UiS4ADL.columns

Index(['timestamp', 'accX[mg]', 'accY[mg]', 'accZ[mg]', 'gyroX[mdps]',
       'gyroY[mdps]', 'gyroZ[mdps]', 'magnX[mG]', 'magnY[mG]', 'magnZ[mG]',
       'fxQ1', 'fxQ2', 'fxQ3', 'fxQ4', 'fxYaw[deg]', 'fxPitch[deg]',
       'fxRoll[deg]', 'fxGravityX[g]', 'fxGravityY[g]', 'fxGravityZ[g]',
       'fxLinAccX[g]', 'fxLinAccY[g]', 'fxLinAccZ[g]', 'fxHeading[deg]', 'adl',
       'session', 'subject', 'fileID'],
      dtype='object')

In [6]:
num_unique_file_ids = df_UiS4ADL['fileID'].nunique()
print(f"Number of unique file IDs: {num_unique_file_ids}")

Number of unique file IDs: 4795


In [7]:
df_UiS4ADL.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9822127 entries, 429317 to 10559907
Data columns (total 28 columns):
 #   Column          Dtype  
---  ------          -----  
 0   timestamp       float64
 1   accX[mg]        float64
 2   accY[mg]        float64
 3   accZ[mg]        float64
 4   gyroX[mdps]     float64
 5   gyroY[mdps]     float64
 6   gyroZ[mdps]     float64
 7   magnX[mG]       float64
 8   magnY[mG]       float64
 9   magnZ[mG]       float64
 10  fxQ1            float64
 11  fxQ2            float64
 12  fxQ3            float64
 13  fxQ4            float64
 14  fxYaw[deg]      float64
 15  fxPitch[deg]    float64
 16  fxRoll[deg]     float64
 17  fxGravityX[g]   float64
 18  fxGravityY[g]   float64
 19  fxGravityZ[g]   float64
 20  fxLinAccX[g]    float64
 21  fxLinAccY[g]    float64
 22  fxLinAccZ[g]    float64
 23  fxHeading[deg]  float64
 24  adl             int64  
 25  session         int64  
 26  subject         int64  
 27  fileID          int64  
dtypes: float64(

In [8]:
# Identify how many NaN values the df dataframe has
nan_count = df_UiS4ADL.isnull().sum().sum()
print(f"Total NaN values in the dataframe: {nan_count}")

# Identify which files contain the empty rows
files_with_nan = df_UiS4ADL.groupby('fileID').apply(lambda group: group.isnull().sum().sum() > 0)
files_with_nan = files_with_nan[files_with_nan].index.tolist()
print(f"Files with NaN values: {files_with_nan}")

# Remove these fileIds from the df
df_UiS4ADL = df_UiS4ADL[~df_UiS4ADL['fileID'].isin(files_with_nan)]

# Verify the rows are removed
print(f"Files with NaN values removed. Remaining rows: {len(df_UiS4ADL)}.")
print(f"New number of unique file IDs: {df_UiS4ADL['fileID'].nunique()}.")

Total NaN values in the dataframe: 183287
Files with NaN values: [4092, 4093]
Files with NaN values removed. Remaining rows: 9814158.
New number of unique file IDs: 4793.


In [9]:
# Count the number of unique subjects in the DataFrame
num_subjects = df_UiS4ADL['subject'].nunique()

print(f"The dataset contains {num_subjects} unique subjects.")

The dataset contains 43 unique subjects.


In [10]:
# Check for duplicate rows
num_dups = df_UiS4ADL.duplicated().sum()

if num_dups > 0:
    print(f"Found {num_dups} duplicates. Removing them...")
    df_UiS4ADL = df_UiS4ADL.drop_duplicates()
else:
    print("No duplicates found.")

# Verify the rows are removed
print(f"Remaining rows: {len(df_UiS4ADL)}")

No duplicates found.
Remaining rows: 9814158


In [11]:
# Group the DataFrame by 'adl', 'subject', and 'session' and count unique 'fileID' values
grouped_file_counts = df_UiS4ADL.groupby(['adl', 'subject', 'session'])['fileID'].nunique()

# Filter and print only the groups that have more than one unique 'fileID' value
multiple_fileID_groups = grouped_file_counts[grouped_file_counts > 1]
if not multiple_fileID_groups.empty:
    print(multiple_fileID_groups)
else:
    print("No groups have more than one unique 'fileID' value.") # every session performed by a subject for a specific ADL is unique

No groups have more than one unique 'fileID' value.


In [12]:
# Describe the DataFrame to show the summary statistics for numerical columns
print("Basic Statistics:")
display(df_UiS4ADL.describe().iloc[1:,0:-4])
print()

Basic Statistics:


,timestamp,accX[mg],accY[mg],accZ[mg],gyroX[mdps],gyroY[mdps],gyroZ[mdps],magnX[mG],magnY[mG],magnZ[mG],...,fxYaw[deg],fxPitch[deg],fxRoll[deg],fxGravityX[g],fxGravityY[g],fxGravityZ[g],fxLinAccX[g],fxLinAccY[g],fxLinAccZ[g],fxHeading[deg]
mean,23823.589599,-31.428892,-279.287428,503.316469,6.771519e+01,-681.848987,98.552465,44.265016,113.676519,8.270501e+01,...,187.348251,-4.193565,-18.141243,0.270436,-0.056761,-0.509976,-0.008805,-0.025338,-0.008644,178.229864
std,25192.968443,513.708140,611.673788,463.672022,7.642712e+04,39369.123743,54898.527164,9243.403013,1273.476978,1.956355e+04,...,106.014525,58.254844,35.171359,0.514111,0.487400,0.478471,0.310362,0.194026,0.223764,102.783641
min,0.000000,-3997.000000,-3997.000000,-3997.000000,-2.061570e+06,-547050.000000,-657790.000000,-13556.000000,-16034.000000,-1.336800e+04,...,0.000000,-180.000000,-90.000000,-1.000000,-1.000000,-1.000000,-4.989000,-4.358000,-4.885000,0.000000
25%,5850.000000,-377.000000,-745.000000,244.000000,-1.708000e+04,-11340.000000,-13510.000000,-318.000000,-100.000000,-4.110000e+02,...,101.160000,-40.450000,-45.380000,-0.089000,-0.401000,-0.853000,-0.079000,-0.072000,-0.066000,90.183000
50%,14560.000000,-25.000000,-371.000000,566.000000,2.100000e+02,-350.000000,-280.000000,-83.000000,122.000000,-2.570000e+02,...,186.950000,-2.810000,-20.460000,0.350000,-0.034000,-0.582000,-0.005000,-0.004000,0.003000,173.401000
75%,32490.000000,268.000000,127.000000,861.000000,1.799000e+04,10150.000000,13720.000000,124.000000,291.000000,-5.700000e+01,...,282.320000,21.630000,5.100000,0.712000,0.232000,-0.284000,0.067000,0.032000,0.050000,268.505000
max,157160.000000,3996.000000,3997.000000,3997.000000,1.712760e+06,621040.000000,530250.000000,829945.000000,111475.000000,1.756986e+06,...,360.000000,180.000000,90.000000,1.000000,1.000000,34.250000,4.873000,4.583000,4.241000,360.000000


## Save the initially cleaned dataset

The 'adl'==0 was removed. Furthermore, there were two files (with IDs 4092 and 4093) that contained NaN values, which lead to 7969 rows being removed as well. No duplicates were found in the dataset.

In [13]:
# Save the DataFrame after initial cleaning
output_file = f"../Data/Processing/combined_UiS4ADL_{fs_UiS4ADL}hz_initially_cleaned.csv"

# Remove the file if it already exists
if os.path.exists(output_file):
    os.remove(output_file)
    print(f"{output_file} has been removed successfully.")
else:
    print(f"{output_file} does not exist.")
    
# Save the cleaned DataFrame to a CSV file
df_UiS4ADL.to_csv(output_file, index=False)
print(f"The dataset was saved to {output_file}")

../Data/Processing/combined_UiS4ADL_100hz_initially_cleaned.csv has been removed successfully.
The dataset was saved to ../Data/Processing/combined_UiS4ADL_100hz_initially_cleaned.csv


# PAAL ADL

In [14]:
fs_PAAL_ADL = 32 # sampling frequency in Hz

path_PAAL_ADL = f"../Data/Processing/PAAL_ADL_{fs_PAAL_ADL}hz_raw.csv"

df_PAAL_ADL = pd.read_csv(path_PAAL_ADL)

In [15]:
df_PAAL_ADL

,timestamp,accX[mg],accY[mg],accZ[mg],adl,session,subject,fileID
0,0.00,-1187.500,125.000,-625.000,7,1,45,1
1,31.25,-859.375,109.375,-531.250,7,1,45,1
2,62.50,-828.125,62.500,-500.000,7,1,45,1
3,93.75,-843.750,78.125,-421.875,7,1,45,1
4,125.00,-1000.000,125.000,-453.125,7,1,45,1
...,...,...,...,...,...,...,...,...
1432168,3500.00,-156.250,-812.500,343.750,18,4,24,6072
1432169,3531.25,109.375,-828.125,515.625,18,4,24,6072
1432170,3562.50,62.500,-750.000,265.625,18,4,24,6072
1432171,3593.75,0.000,-703.125,218.750,18,4,24,6072


## Basic data cleaning and overview

In [16]:
print(f'The dataset contains {df_PAAL_ADL["adl"].nunique()} unique ADL classes.')

The dataset contains 24 unique ADL classes.


In [17]:
df_PAAL_ADL.columns

Index(['timestamp', 'accX[mg]', 'accY[mg]', 'accZ[mg]', 'adl', 'session',
       'subject', 'fileID'],
      dtype='object')

In [18]:
num_unique_file_ids = df_PAAL_ADL['fileID'].nunique()
print(f"Number of unique file IDs: {num_unique_file_ids}")

Number of unique file IDs: 6072


In [19]:
df_PAAL_ADL.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1432173 entries, 0 to 1432172
Data columns (total 8 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   timestamp  1432173 non-null  float64
 1   accX[mg]   1432173 non-null  float64
 2   accY[mg]   1432173 non-null  float64
 3   accZ[mg]   1432173 non-null  float64
 4   adl        1432173 non-null  int64  
 5   session    1432173 non-null  int64  
 6   subject    1432173 non-null  int64  
 7   fileID     1432173 non-null  int64  
dtypes: float64(4), int64(4)
memory usage: 87.4 MB


In [20]:
# Identify how many NaN values the df dataframe has
nan_count = df_PAAL_ADL.isnull().sum().sum()
print(f"Total NaN values in the dataframe: {nan_count}")

Total NaN values in the dataframe: 0


In [21]:
# Count the number of unique subjects in the DataFrame
num_subjects = df_PAAL_ADL['subject'].nunique()

print(f"The dataset contains {num_subjects} unique subjects.")

The dataset contains 52 unique subjects.


In [22]:
# Check for duplicate rows
num_dups = df_PAAL_ADL.duplicated().sum()

if num_dups > 0:
    print(f"Found {num_dups} duplicates. Removing them...")
    df_PAAL_ADL = df_PAAL_ADL.drop_duplicates()
else:
    print("No duplicates found.")

# Verify the rows are removed
print(f"Remaining rows: {len(df_PAAL_ADL)}")

No duplicates found.
Remaining rows: 1432173


In [23]:
# Group the DataFrame by 'adl', 'subject', and 'session' and count unique 'fileID' values
grouped_file_counts = df_PAAL_ADL.groupby(['adl', 'subject', 'session'])['fileID'].nunique()

# Filter and print only the groups that have more than one unique 'fileID' value
multiple_fileID_groups = grouped_file_counts[grouped_file_counts > 1]
if not multiple_fileID_groups.empty:
    print(multiple_fileID_groups)
else:
    print("No groups have more than one unique 'fileID' value.") # every session performed by a subject for a specific ADL is unique

No groups have more than one unique 'fileID' value.


In [24]:
# Describe the DataFrame to show the summary statistics for numerical columns
print("Basic Statistics:")
display(df_PAAL_ADL.describe().iloc[1:,0:-4])
print()

Basic Statistics:


,timestamp,accX[mg],accY[mg],accZ[mg]
mean,8562.432271,124.888467,-2.509495,396.040176
std,13178.158913,657.546079,551.565271,504.890856
min,0.000000,-2000.000000,-2000.000000,-2000.000000
25%,2000.000000,-406.250000,-390.625000,109.375000
50%,5562.500000,218.750000,0.000000,500.000000
75%,11531.250000,640.625000,375.000000,765.625000
max,254593.750000,1984.375000,1984.375000,1984.375000


## Save the initially cleaned dataset

The The PAAL_ADL dataset contained no NaN values and no duplicate rows. It includes 1,279 more files than the UiS4ADL dataset, and its data collection involved 9 additional subjects.

Since no modifications were made to the file, no new files will be saved.